xrdp/xrdp_mm.c

```diff
@@ -1008,194 +1021,343 @@ dynamic_monitor_data_first(intptr_t id, int chan_id, char *data, int bytes,
     return 0;
 }
 
+/******************************************************************************/
+int
+advance_resize_state_machine(struct xrdp_mm *mm,
+                             struct display_control_monitor_layout_data *description,
+                             enum display_resize_state new_state)
+{
+    LOG_DEVEL(LOG_LEVEL_INFO,
+              "advance_resize_state_machine:"
+              " Processing resize to: %d x %d."
+              " Advancing state from %s to %s."
+              " Previous state took %d MS.",
+              description->description.session_width,
+              description->description.session_height,
+              XRDP_DISPLAY_RESIZE_STATE_TO_STR(description->state),
+              XRDP_DISPLAY_RESIZE_STATE_TO_STR(new_state),
+              g_time3() - description->last_state_update_timestamp);
+    description->state = new_state;
+    description->last_state_update_timestamp = g_time3();
+    g_set_wait_obj(mm->resize_ready);
+    return 0;
+}
+
 /******************************************************************************/
 static int
-process_dynamic_monitor_description(struct xrdp_wm *wm,
-                                    struct display_size_description *description)
+dynamic_monitor_data(intptr_t id, int chan_id, char *data, int bytes)
 {
     int error = 0;
-    struct xrdp_mm *mm = wm->mm;
-    struct xrdp_mod *module = mm->mod;
+    struct stream ls;
+    struct stream *s;
+    int msg_type;
+    int msg_length;
+    struct xrdp_process *pro;
+    struct xrdp_wm *wm;
+    int monitor_layout_size;
+    struct display_size_description *display_size_data;
 
-    LOG_DEVEL(LOG_LEVEL_TRACE, "process_dynamic_monitor_description:");
+    LOG_DEVEL(LOG_LEVEL_TRACE, "dynamic_monitor_data:");
+    pro = (struct xrdp_process *) id;
+    wm = pro->wm;
 
     if (wm->client_info->suppress_output == 1)
     {
-        LOG(LOG_LEVEL_ERROR,
-            "process_dynamic_monitor_description:"
-            " Not allowing resize. Suppress output is active.");
-        return 0;
+        LOG(LOG_LEVEL_INFO, "dynamic_monitor_data: Not allowing resize."
+            " Suppress output is active.");
+        return error;
     }
-    if (description == NULL)
+
+    g_memset(&ls, 0, sizeof(ls));
+    ls.data = data;
+    ls.p = ls.data;
+    ls.size = bytes;
+    ls.end = ls.data + bytes;
+    s = &ls;
+    in_uint32_le(s, msg_type);
+    in_uint32_le(s, msg_length);
+    LOG(LOG_LEVEL_DEBUG, "dynamic_monitor_data: msg_type %d msg_length %d",
+        msg_type, msg_length);
+
+    if (msg_type != DISPLAYCONTROL_PDU_TYPE_MONITOR_LAYOUT)
     {
-        LOG_DEVEL(LOG_LEVEL_ERROR,
-                  "process_dynamic_monitor_description:"
-                  " description is null.");
         return 0;
```

Comment:
Are there other message types that this function is expected to process but is not supported?
Can you add a LOG message for the message types that are ignored here.

Reply:
Added this code block

```
 // Unsupported message types
        switch (msg_type)
        {
            case DISPLAYCONTROL_PDU_TYPE_CAPS:
                LOG(LOG_LEVEL_ERROR, "dynamic_monitor_data: Received"
                    " DISPLAYCONTROL_PDU_TYPE_CAPS. Per MS-RDPEDISP 2.2.2.1,"
                    " this is a server-to-client message, client should never"
                    " send this. Ignoring message");
                break;
            default:
                LOG(LOG_LEVEL_ERROR, "dynamic_monitor_data: Received neither"
                    " nor DISPLAYCONTROL_PDU_TYPE_CAPS. Ignoring message.");
                break;
        }
```

Hope that settles this concern.